contributivity_measures.py

```diff
@@ -176,4 +199,980 @@ def not_twice_characteristic(permut):
+        v_max = 0
+        while (
+            t < 100 or t < 4 * q ** 2 * v_max / (sv_accuracy) ** 2
+        ):  # Check if the length of the confidence interval  is below the value of sv_accuracy*characteristic_all_node
+            t += 1
+            if t == 1:
+                contributions = np.array([np.zeros(n)])
+            else:
+                contributions = np.vstack((contributions, np.zeros(n)))
+
+            permutation = np.random.permutation(n)  # Store the current permutation
+            char_nodelists = np.zeros(
+                n + 1
+            )  # Store the characteristic function on each ensemble built with the first elements of the permutation
+            char_nodelists[-1] = characteristic_all_node
+            first = True
+            for j in range(n):
+                # here we suppose the characteristic function is 0 for the empty set
+                if abs(characteristic_all_node - char_nodelists[j]) < truncation:
+                    if first:
+                        size_of_rest = 0
+                        for i in range(j, n):
+                            size_of_rest += len(preprocessed_node_list[i].y_train)
+                        first = False
+                    size_of_S = len(preprocessed_node_list[j].y_train)
+                    char_nodelists[j + 1] = (
+                        char_nodelists[j]
+                        + size_of_S / size_of_rest * characteristic_all_node
+                    )
+                else:
+                    char_nodelists[j + 1] = not_twice_characteristic(
+                        permutation[: j + 1]
+                    )
+                contributions[-1][permutation[j]] = (
+                    char_nodelists[j + 1] - char_nodelists[j]
+                )
+            v_max = np.max(np.var(contributions, axis=0))
+        shap = np.mean(contributions, axis=0)
+
+    return {
+        "sv": shap,
+        "std_sv": np.std(contributions, axis=0) / np.sqrt(t - 1),
+        "prop": shap / np.sum(shap),
+        "computed_val": char_value_dict,
+        "fit_count": the_scenario.fit_count - nb_fit_begining,
+    }
+
+
+# %% compute Shapley values with the importance sampling method
+
+
+def IS_lin(the_scenario, sv_accuracy=0.01, alpha=0.95):
+    """Return the vector of approximated Shapley value corresponding to a list of node and a characteristic function using the importance sampling method and a linear interpolation model."""
+
+    nb_fit_begining = the_scenario.fit_count
+    preprocessed_node_list = the_scenario.node_list
+    n = len(preprocessed_node_list)
+
+    # We store the value of the characteristic function in order to avoid recomputing it twice
+    char_value_dict = {(): 0}  # the dictionary that will countain the values
+
+    # Return the characteristic function of the nodelist associated to the ensemble permut, without recomputing it if it was already computed
+    def not_twice_characteristic(permut):
```

Comment:
As `not_twice_characteristic(permut)` is used in multiple contributivity measurement methods, it would be great to make it a independent function.

Reply:
I agree. But I need to store the values of the characteristic function that were already computed somewhere. So far it is stored in a local variable in each method, that's why I repeated the not_twice characteristic function inside each methods so it can access this local variable. 
I was thinking to put all the contributivity measures as class method of  the contributivity class and to store the computed values (and the counts) as variables attributes of the contributivity instance. Do think it is a good idea?

Reply:
I don't have enough background in Python to answer this question. @natct10, @RomainGoussault, would you guys have an educated opinion on that one?

Reply:
Hello, j'essaie de suivre ce qui se passe dans ces méthodes, mais du coup, pourquoi ne pas faire une méthode plus générique à laquelle on passerait tous les paramètres qui vont bien et pas juste le `permut` ? En gardant les conditions "if already computed" / "else"

Reply:
We could do that, yes.

Reply:
Classmethod can be an alternative otherwise